service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java

```diff
@@ -619,71 +624,97 @@ public HonoConnection commandConsumerConnection() {
-                            final Duration lifespan,
-                            final SpanContext context) {
-                        return deviceConnectionClient.setCommandHandlingAdapterInstance(tenant, deviceId, adapterInstanceId, lifespan, context);
-                    }
-
-                    @Override
-                    public Future<Void> removeCommandHandlingAdapterInstance(
-                            final String tenant,
-                            final String deviceId,
-                            final String adapterInstanceId,
-                            final SpanContext context) {
-                        return deviceConnectionClient.removeCommandHandlingAdapterInstance(tenant, deviceId, adapterInstanceId, context);
-                    }
-                });
-
-        return commandConsumerFactory;
+            final CommandRouterClient commandRouterClient) {
+
+        if (commandRouterClient instanceof DeviceConnectionClient) {
+
+            final DeviceConnectionClient deviceConnectionClient = (DeviceConnectionClient) commandRouterClient;
+            final CommandTargetMapper commandTargetMapper = CommandTargetMapper.create(getTracer());
+
+            commandTargetMapper.initialize(new CommandTargetMapperContext() {
+
+                @Override
+                public Future<List<String>> getViaGateways(
+                        final String tenant,
+                        final String deviceId,
+                        final SpanContext context) {
+
+                    Objects.requireNonNull(tenant);
+                    Objects.requireNonNull(deviceId);
+
+                    return registrationClient.assertRegistration(tenant, deviceId, null, context)
+                            .map(RegistrationAssertion::getAuthorizedGateways);
+                }
+
+                @Override
+                public Future<JsonObject> getCommandHandlingAdapterInstances(
+                        final String tenant,
+                        final String deviceId,
+                        final List<String> viaGateways,
+                        final SpanContext context) {
+
+                    Objects.requireNonNull(tenant);
+                    Objects.requireNonNull(deviceId);
+                    Objects.requireNonNull(viaGateways);
+
+                    return deviceConnectionClient.getCommandHandlingAdapterInstances(
+                            tenant, deviceId, viaGateways, context);
+                }
+            });
+
+            return new ProtonBasedCommandConsumerFactory(
+                    commandConsumerConnection(),
+                    samplerFactory,
+                    adapterProperties,
+                    commandTargetMapper,
+                    new CommandHandlingAdapterInfoAccess() {
+
+                        @Override
+                        public Future<Void> setCommandHandlingAdapterInstance(
+                                final String tenant,
```

Comment:
The creation of the `CommandHandlingAdapterInfoAccess` object here and in the Quarkus `AbstractProtocolAdapterApplication` can be moved to the `ProtonBasedCommandConsumerFactory` constructor, passing along the  `DeviceConnectionClient` object instead of `CommandHandlingAdapterInfoAccess`.

Reply:
> But it makes the factory (again) dependent on a particular other client which we already saw to be problematic.

As far as I remember, such a case was problematic when creating the factory in a bean method directly, where e.g. the used registration client bean was created via a `@Scope("prototype")` bean definition and there were therefore issues when requiring the very same registration client bean instance in the adapter instance that used the above factory bean.

But, with the `setCollaborators` method, from which the above `commandConsumerFactory()` method is called, we have already decoupled that. I don't see that another layer of decoupling is needed here, seeing that usage of the `DeviceRegistrationClient` is the primary way to create the factory. Therefore I think creating an additional `ProtonBasedCommandConsumerFactory` constructor with a `DeviceRegistrationClient` parameter would be a good way to remove some clutter and duplicate code from the `AbstractAdapterConfig` and `AbstractProtocolAdapterApplication` classes.

Reply:
Ok, I'll do that.